src/rules/export.js

```diff
@@ -45,6 +45,41 @@ function isTypescriptFunctionOverloads(nodes) {
   );
 }
 
+/**
+ * Detect merging Namespaces with Classes, Functions, or Enums like:
+ * ```ts
+ * export class Foo { }
+ * export namespace Foo { }
+ * ```
+ * @param {Set<Object>} nodes
+ * @returns {boolean}
+ */
+function isTypescriptNamespaceMerging(nodes) {
+  const types = new Set(Array.from(nodes, node => node.parent.type));
+  const noNamespaceNodes = Array.from(nodes).filter((node) => node.parent.type !== 'TSModuleDeclaration');
+
+  const isMerging = (
+    types.has('TSModuleDeclaration') &&
+    (
+      types.size === 1 ||
+      // Merging with functions
+      (types.size === 2 && (types.has('FunctionDeclaration') || types.has('TSDeclareFunction'))) ||
+      (types.size === 3 && types.has('FunctionDeclaration') && types.has('TSDeclareFunction')) ||
+      // Merging with classes or enums
+      (types.size === 2 && (types.has('ClassDeclaration') || types.has('TSEnumDeclaration')) && noNamespaceNodes.length === 1)
+    )
+  );
+
+  if (!isMerging && types.has('TSModuleDeclaration') && (types.has('TSEnumDeclaration') || types.has('ClassDeclaration') || types.has('FunctionDeclaration') || types.has('TSDeclareFunction'))) {
+    // Remove namespace nodes to error only on other types
+    Array.from(nodes).forEach(node => {
+      if (node.parent.type === 'TSModuleDeclaration') nodes.delete(node);
```

Comment:
mutating an argument seems very undesirable

Reply:
I was a little hesitant about the best way to do it, but I found a different approach implemented on [`d7acbba` (#2375)](https://github.com/import-js/eslint-plugin-import/pull/2375/commits/d7acbbac2619acf5d104fc12f393252a168ed8eb).